pkg/webui/console/containers/device-onboarding-form/device-type-form-section/index.js

```diff
@@ -38,61 +37,74 @@ const initialValues = merge(
   repositoryInitialValues,
 )
 
-const DeviceTypeFormSection = props => {
-  const { getRegistrationTemplate, appId } = props
+const DeviceTypeFormSection = () => {
   const {
-    values: { _inputMethod, _isClaiming },
+    values: { version_ids, frequency_plan_id, lorawan_version, lorawan_phy_version, _inputMethod },
     resetForm,
+    initialValues,
   } = useFormContext()
+
+  const isPristineForm =
+    (!hasCompletedDeviceRepositorySelection(version_ids) ||
+      hasSelectedDeviceRepositoryOther(version_ids)) &&
+    !Boolean(frequency_plan_id) &&
+    !Boolean(lorawan_phy_version) &&
+    !Boolean(lorawan_version)
+
   const [showModal, setShowModal] = useState(false)
 
-  const handleMethodChange = React.useCallback(() => {
-    setShowModal(true)
-  }, [])
+  const handleMethodChange = React.useCallback(
+    value => {
+      if (!isPristineForm) {
+        setShowModal(true)
+      } else {
+        resetForm({ values: { ...initialValues, _inputMethod: value } })
+      }
+    },
+    [initialValues, isPristineForm, resetForm],
+  )
 
-  const handleComplete = React.useCallback(() => {
-    setShowModal(false)
-    resetForm({ values: { ...initialValues, _inputMethod } })
-  }, [resetForm, _inputMethod])
+  const handleMethodModalComplete = React.useCallback(
+    approved => {
+      setShowModal(false)
+      if (approved) {
+        resetForm({
+          values: {
+            ...initialValues,
+            _inputMethod: _inputMethod === 'device-repository' ? 'manual' : 'device-repository',
+          },
+        })
+      }
+    },
+    [_inputMethod, initialValues, resetForm],
+  )
 
   return (
     <>
-      {showModal && _isClaiming !== undefined && (
+      {showModal && (
         <Modal
           buttonMessage={m.changeDeviceTypeButton}
           message={m.changeDeviceType}
           title={m.changeDeviceTypeButton}
-          onComplete={handleComplete}
+          onComplete={handleMethodModalComplete}
         />
       )}
       <Form.SubTitle title={m.endDeviceType} />
       <Form.Field
         title={m.inputMethod}
-        name="_inputMethod"
         component={Radio.Group}
         onChange={handleMethodChange}
+        value={_inputMethod}
+        name="__inputMethod"
+        connectedFields={['_inputMethod']}
       >
```

Comment:
We were originally only using the `onChange` handler to control the modal here. The problem with this is that the `onChange` handler only gets executed after the value was already changed in the form. So regardless of the modals decision, the value would already be changed regardless.
To fix this, I added an intermediate field value `__inputMethod` which is not really used but allows us to control the actual value `_inputMethod` based on the result of the modal.

I also changed the conditions for when the modal would show, since it should only show when some progress in the form was actually already made which would be lost.